wpilibj/src/test/java/edu/wpi/first/wpilibj/event/BooleanEventTest.java

```diff
@@ -58,15 +59,89 @@ void testBinaryCompositionLoopSemantics() {
 
     assertEquals(1, counter1.get());
     assertEquals(1, counter2.get());
+
+    bool2.set(false);
+    loop1.poll();
+
+    assertEquals(2, counter1.get());
+    assertEquals(1, counter2.get());
+
+    loop2.poll();
+
+    assertEquals(2, counter1.get());
+    assertEquals(1, counter2.get());
+
+    loop1.poll();
+
+    assertEquals(2, counter1.get());
+    assertEquals(1, counter2.get());
+
+    bool2.set(true);
+    loop2.poll();
+
+    assertEquals(2, counter1.get());
+    assertEquals(1, counter2.get());
+
+    loop1.poll();
+
+    assertEquals(3, counter1.get());
+    assertEquals(1, counter2.get());
+
+    bool1.set(false);
+    loop2.poll();
+
+    assertEquals(3, counter1.get());
+    assertEquals(1, counter2.get());
+
+    loop1.poll();
+
+    assertEquals(3, counter1.get());
+    assertEquals(1, counter2.get());
+
+    loop2.poll();
+
+    assertEquals(3, counter1.get());
+    assertEquals(1, counter2.get());
+  }
+
+  /**
+   * When a BooleanEvent is constructed, an action is bound to the event loop to update an internal
+   * state variable. This state variable is checked during loop polls to determine whether or not to
+   * execute an action. If a condition is changed during the loop poll but before the state variable
+   * gets updated, the changed condition is used for the state variable.
+   */
+  @Test
+  void testEventConstructionOrdering() {
```

Comment:
I think I understand what's being tested here, but I don't think this is a robust way of testing it -- the link between construction order and whether the callback is invoked or not is rather far/distant.

Perhaps a single counter that each of the three lambdas (`event1`'s condition, `event2`'s condition, and the lambda bound to their composition) assigns to a different variable and increments (ie the counter is read with `getAndIncrement`/`incrementAndGet`), and then assert that the corresponding variables have the correct values?  It can also be repeated multiple loop polls to ensure that the events are polled in the correct order. Both condition lambdas would return true.

Reply:
I'm not sure what the best way to express this test is. What happens is that `event1` (the lambda which changes the bool) is constructed first, followed by `event2` (bool2::get) and then the composed event. This means when the loop is polled, event1's condition is always called first, then event2's condition, and finally the composed event. This causes `event2` to never be true, since event1's condition always runs first and sets `bool2` to false, so when the loop checks `event2`, it always sees false. Switching the order should cause the composed event to execute.

Fundamentally, the behavior that's being tested is that `m_state` is set when the loop calls the event condition, not at the start of the loop poll (which for most use cases actually doesn't matter). This tests checks that behavior by changing the event condition right before the loop calls it.